internal/batch/batch_processor.go

```diff
@@ -101,103 +130,249 @@ func (bw *batchWork) estimateSize() int64 {
 	return sizeEstimate
 }
 
-// The assemblyLoop accepts work into the pipe as quickly as possible.
-// It dispatches work asynchronously to the persistenceLoop, which is responsible for
-// calling back each piece of work once persisted into a batch
-// (doesn't wait until that batch is sealed/dispatched).
-// The assemblyLoop seals batches when they are full, or timeout.
+func (bp *batchProcessor) status() *ProcessorStatus {
+	bp.statusMux.Lock()
+	defer bp.statusMux.Unlock()
+	return &ProcessorStatus{
+		Dispatcher: bp.conf.dispatcherName,
+		Name:       bp.conf.name,
+		Status:     bp.flushStatus, // copy
+	}
+}
+
+func (bp *batchProcessor) newAssembly(initalWork ...*batchWork) {
+	bp.assemblyID = fftypes.NewUUID()
+	bp.assemblyQueue = append([]*batchWork{}, initalWork...)
+	bp.assemblyQueueBytes = batchSizeEstimateBase
+}
+
+// addWork adds the work to the assemblyQueue, and calculates if we have overflowed with this work.
+// We check for duplicates, and add the work in sequence order.
+// This helps in the case for parallel REST APIs all committing to the DB at a similar time.
+// With a sufficient batch size and batch timeout, the batch will still dispatch the messages
+// in DB sequence order (although this is not guaranteed).
+func (bp *batchProcessor) addWork(newWork *batchWork) (full, overflow bool) {
```

Comment:
I would probably make these `bool` returns into an enum but you know how I roll 🤓

Reply:
... yeah... in my opinion with the code as it's structure they are independent things, so we'd end up with 4 states.
So I'm going to stick with the bools here unless you feel strongly.